homeassistant/components/lmstudio/entity.py

```diff
@@ -0,0 +1,237 @@
+    chat_log: conversation.ChatLog,
+    result: AsyncGenerator[ChatCompletionChunk],
+) -> AsyncGenerator[conversation.AssistantContentDeltaDict]:
+    """Transform OpenAI stream to Home Assistant format."""
+    async for chunk in result:
+        if not chunk.choices:
+            continue
+
+        choice = chunk.choices[0]
+
+        if choice.delta.content:
+            yield conversation.AssistantContentDeltaDict(
+                content=choice.delta.content,
+            )
+
+        if choice.delta.tool_calls:
+            for tool_call in choice.delta.tool_calls:
+                if tool_call.function and tool_call.function.name:
+                    yield conversation.AssistantContentDeltaDict(
+                        tool_calls=[
+                            llm.ToolInput(
+                                tool_name=tool_call.function.name,
+                                tool_args=json.loads(
+                                    tool_call.function.arguments or "{}"
+                                ),
+                            )
+                        ]
+                    )
+
+
+class LMStudioBaseLLMEntity(Entity):
+    """Base LM Studio LLM entity."""
+
+    _attr_should_poll = False
+
+    def __init__(self, entry: LMStudioConfigEntry, subentry: ConfigSubentry) -> None:
+        """Initialize the entity."""
+        self.entry = entry
+        self.subentry = subentry
+        self._attr_unique_id = f"{entry.entry_id}-{subentry.subentry_id}"
+        self._attr_name = subentry.title
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, entry.entry_id)},
+            name=f"LM Studio ({entry.data[CONF_BASE_URL]})",
+            manufacturer="LM Studio",
+            model="Local LLM Server",
+            configuration_url=entry.data[CONF_BASE_URL],
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        return self.entry.runtime_data is not None
+
+    async def _async_handle_chat_log(self, chat_log: conversation.ChatLog) -> None:
+        """Handle the chat log and generate response."""
+        client = self.entry.runtime_data
+        options = {**self.entry.data, **self.subentry.data}
+
+        model = options.get(CONF_MODEL, "")
+        if model is None or model == "":
+            # If no model specified, use the first available model
+            try:
+                models_response = await client.with_options(timeout=10.0).models.list()
+                if models_response.data:
+                    model = models_response.data[0].id
+                else:
+                    raise HomeAssistantError("No models available on LM Studio server")
+            except openai.OpenAIError as err:
+                raise HomeAssistantError(f"Failed to get models: {err}") from err
+
+        # Build the messages from the chat log
+        messages: list[dict[str, Any]] = []
+
+        for message in chat_log.content:
```

Comment:
You might consider switching this to `match/case` (py310+) and/or create a helper function for this. I think a match/case (within a helper) might be the cleanest option. (Opinionated review, the combined for/if/for/if have a high complexity rating)